src/ifcopenshell-python/ifcopenshell/type_checker.py

```diff
@@ -0,0 +1,149 @@
+        return int(''.join(token.value for token in items))
+
+    def day(self, items):
+        return int(''.join(token.value for token in items))
+
+    def hour(self, items):
+        return int(''.join(token.value for token in items))
+
+    def minute(self, items):
+        return int(''.join(token.value for token in items))
+
+    def second(self, items):
+        return ''.join(token.value for token in items)
+
+    def millisecond(self, items):
+        if items:
+            return ''.join(token.value for token in items)
+        else:
+            return None
+
+    def time_value(self, items):
+
+        year, month, day, hour, minute, second, millisecond = items
+
+        return {
+            "year": year,
+            "month": month,
+            "day": day,
+            "hour": hour,
+            "minute": minute,
+            "second": float(second + f".{millisecond if millisecond else ''}")
+        }
+
+
+def parse_datetime(datetime_value):
+    try:
+        tree = parser.parse(datetime_value)
+        transformer = DatetimeTransformer()
+        parsed_datetime = transformer.transform(tree)
+        return {"status": 1, "msg": parsed_datetime}
+    except Exception as e:
+        return {"status": -1, "msg": e}
+
+
+def validate_datetime(parsed_datetime):
+    if parsed_datetime == -1:
+        return "Syntax error"
+    else:
+        # Year
+        if parsed_datetime["year"] == "0000":
+            return "invalid year value"
+        if parsed_datetime["year"][0] != "-" and len(parsed_datetime["year"]) > 4:
+            return "invalid year value"
+
+        # Month
+        if not (parsed_datetime["month"] > 0 and parsed_datetime["month"] < 13):
+            return "invalid month value"
+
+        # Day
+        if not (parsed_datetime["day"] > 0 and parsed_datetime["day"] < 32):
+            return "invalid day value"
+
+        # Check February and Leap year case
+        if parsed_datetime["month"] == 2:
+            if int(parsed_datetime["year"]) % 4 == 0:
+                if not (parsed_datetime["day"] < 30):
+                    return "invalid day value"
+            else:
+                if not (parsed_datetime["day"] < 29):
+                    return "invalid day value"
+
+        # Hour
+        if not (parsed_datetime["hour"] > -1 and parsed_datetime["hour"] < 24):
+            return "invalid hour value"
+
+        # Minute
+        if not (parsed_datetime["minute"] > -1 and parsed_datetime["minute"] < 60):
+            return "invalid minute value"
+
+        # Second
+        if not (parsed_datetime["second"] > -1 and parsed_datetime["second"] < 60):
+            return "invalid second value"
+
+        return 1
+
+
+def validate(d):
+    parsing = parse_datetime(d)
+    if parsing["status"] != -1:
+        if validate_datetime(parsing["msg"]) != 1:
+            print(validate_datetime(parsing["msg"]))
+    else:
+        print("syntax error")
+
+
+if __name__ == '__main__':
```

Comment:
Please add these as a `def test_datetime()` in https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/test/test_validate.py

Reply:
Also we should add some test cases of actual IFC file content to check whether the values of such types are correctly discovered. For those you can follow the same convention as here https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.7.0/src/ifcopenshell-python/test/fixtures/validate

E.g create an IfcPropertyListValue with multiple IfcDateTime values. IfcDateTime is IfcSimpleValue is IfcValue.

Reply:
test files created https://github.com/IfcOpenShell/IfcOpenShell/pull/3864/commits/d0abe8fa7afc77fc5e8bfd64c8aac91d56383e2f